keep reactive computations synchronous with the render phase

menelike opened the pull request:
- rewrite useTracker in order to stay fully consistent with current withTracker behavior
- compare deps in order to keep API consistency to React.useEffect()

fixes https://github.com/meteor/react-packages/pull/262

Commit:
- rewrite useTracker in order to stay fully consistent with current withTracker behavior
- compare deps in order to keep API consistency to React.useEffect()

Commit:
- withTracker should always recompute on re-render so deps for useTracker() can be omitted
- also React.memo() already has a check for prop change so there is no need to check for changed deps again

Commit:
update Readme to reflect omitted deps behavior

Commit:
fix code comment

Commit:
get rid of Math.random(), wasn't needed at all

Commit:
don't handle Meteor.isServer as it's already been taken care of in exports

CaptainN commented on packages/react-meteor-data/useTracker.js:
```diff
@@ -28,41 +28,101 @@ function checkCursor(data) {
+          // use Math.random() to trigger a state change to enforce a re-render
+          // Calling forceUpdate() triggers componentWillUpdate which
+          // calls the reactive function and re-renders the component.
+          forceUpdate(Math.random());
```
The note I made about the use of `Math.random` was actually meant for here - it's super unlikely, but possible that `Math.random()` produces the same value as is already stored, which would mean it wouldn't trigger update. An incrementing value would work just as well here, and would prevent that from happening.

```js
const [forceValue, forceUpdate] = useState();
...
forceUpdate(forceValue + 1);
```

If there is something about react concurrent that might throw a wrench in the gears for that, then please disregard.

menelike replied:
Thanks for the catch, I totally missed that. A simple counter is good enough, https://github.com/yched/react-packages/pull/1/commits/c3803b96cf2eb843c8df549753586d398352f201 should do the job. (initially I wanted to use the id of the computation which is a counter itself the becomes incremented per computation, but since it is a private variable I used an own counter)

Commit:
replace Math.random() when enforcing an update

Commit:
- align areHookInputsEqual with

https://github.com/facebook/react/blob/d77d12510b1a1c37484d771a323e0a02cbeb9ba7/packages/react-reconciler/src/ReactFiberHooks.js#L231
https://github.com/facebook/react/blob/d77d12510b1a1c37484d771a323e0a02cbeb9ba7/packages/react-reconciler/src/ReactFiberHooks.js#L318-L329
https://github.com/facebook/react/blob/d77d12510b1a1c37484d771a323e0a02cbeb9ba7/packages/react-reconciler/src/ReactFiberHooks.js#L878

- update comments

Commit:
warn if dep is not an array when Meteor isDevelopment

Commit:
fix prevDeps isArray check

Commit:
warn if initial deps is not an array

CaptainN committed:
Retain synchronous render behavior for firstRun (including after deps change), but allow async and computation reuse after.

CaptainN committed:
Fix eslint errors

CaptainN committed:
Merge branch 'hooks' into half-duplex

# Conflicts:
#	packages/react-meteor-data/useTracker.js

CaptainN committed:
disambiguate the disposition of previous computation - this works the same as before, but is easier to read

menelike committed:
Merge pull request #1 from CaptainN/half-duplex

Half duplex

CaptainN committed:
Merge branch 'half-duplex' into hooks

CaptainN committed:
Use 1 useRef instead of multiple

CaptainN committed:
If reactiveFn will run synchrously next render, don't bother running it asynchronously

CaptainN committed:
Dispose of the computation early, if the deps are falsy on meteor reactive changes

CaptainN committed:
previousDeps will always equal deps at this point, so just check previousDeps

menelike committed:
Merge pull request #3 from CaptainN/hooks

Hooks

CaptainN commented:
This is good to go @yched. It retains backward compatibility with `withTracker` (or if deps are not specified/falsy), retains the ability to reuse the computation based on deps, and performs asynchronous reactivity if deps are specified. Should be ready to merge!

CaptainN commented:
Any thoughts on this @yched ?

yched commented:
Catching up :-)

Since you opened https://github.com/meteor/react-packages/pull/271 on the main repo, I guess this PR here can be closed now, right ? (let's not fork the discussion more than it currently is ;-) )

menelike commented:
superseded by https://github.com/meteor/react-packages/pull/271
